### ISSUE TITLE
Querying Run Properties

### ISSUE THREAD

**Issue by jcasse** [Netflix/metaflow#226]
To obtain the value of the properties such as run.finished or run.successful, metaflow queries AWS (I think S3?).

If you have a lot of runs and you want to check whether they have finished, then that requires a query to AWS for each run property. Most likely the majority of those runs have already finished.

Would it be possible to periodically update RDS so that when a run object is created and its metadata fetched, the value of the property is already available? Only when the property value is not "True", is querying AWS warranted to get an updated value.

**Comment by savingoyal** [comment]
Hi @jcasse, the value of the properties `run.finished` as well as `run.successful` doesn't query the datastore (`S3` or `local`) and is already being fetched directly from the RDS instance. It's actually an interesting problem to figure out when the run has been `lost` and we will be adding a heart-beat functionality in the short order so that we can better signal that staus.

**Comment by jcasse** [comment]
Hi @savingoyal, whenever I execute `run.finished`, it does go to AWS. The first sign that hinted to that is that it takes some time to get the value. I also verified it using Wireshark.

**Comment by jwang01** [comment]
@savingoyal So RDS Run_V3 table does not store the finished or successful info, I believe the file on S3 like 0.runtime.json (shown below) has the info, so it needs to pull the info from S3 to complete the query.
```
{"return_code": 0, "killed": false, "success": true}
```

**Comment by savingoyal** [comment]
@jcasse @jwang01 Currently we don't have a way of identifying failed tasks (hence my comment around adding heartbeats above). For `finished` property, we actually create a `_task_ok` artifact when the task succeeds and [use the creation time of this artifact as the finished timestamp of the task](https://github.com/Netflix/metaflow/blob/master/metaflow/client/core.py#L935). We should definitely look into why this RDS query is slow.

**Comment by jwang01** [comment]
@savingoyal I think db query is not slow, but it's slow to check the artifact from s3.

**Comment by jcasse** [comment]
Scenario:
```
flow = metaflow.Flow(f'{flow_name}')
runs = flow.runs()  # Gets data from RDS (this happens quickly, as expected)
for run in runs:
    created_at = run.created_at  # Value already locally
    finished = run.finished  # Goes to S3 every time invoked to get updated value
```
When you have many runs, say, 30, it will take over 30 seconds!!
I understand that while `run.finished` is `False`, we must go to `S3 `to obtain the updated value.
But, most runs in the list have already finished. So, what I ask is if the `run.finished` information could behave like `run.created_at` once it becomes `True`. Just as `run.created_at`, once `run.finished` is `True`, it will not change, so no need to go to `S3`.

**Comment by savingoyal** [comment]
Ah yes, you are indeed correct. `finished_at` only hits RDS while `finished` hits S3. We should modify the behavior of `finished` to return `True` when `finished_at` is not `None`.

**Comment by savingoyal** [comment]
Turns out `finished` isn't necessarily `True` when `finished_at` is not `None` since we consider tasks with unhandled exceptions as [`unfinished`](https://github.com/Netflix/metaflow/blob/master/metaflow/task.py#L430). The proposed change above will break the current semantics. I will introduce a `running` call that will help with the semantics you are looking for in the near term.

**Comment by jcasse** [comment]
@savingoyal thanks for looking into this. Having the `running` semantics is always good. However, if `running` is a call that queries S3, then the `for` loop in the example above will still take over 30 seconds. I think that if the `finished` state is stored in RDS and periodically updated, then we can avoid having to query S3 for an update by first checking if `finished` is already `True`. The same applies to `successful`. And, the `running` state.

**Comment by savingoyal** [comment]
@jcasse Yes, we will provide a fast path for `running` which doesn't rely on the datastore (S3) and goes directly against RDS.
